File: benchmarks/javascript/run_accuracy.mjs
/**
 * JS Accuracy Benchmark — prompt-sanitizer vs OpenRedaction
 * ==========================================================
 *
 * Usage:
 *   npm install
 *   node run_accuracy.mjs
 *
 * Metric: Value-overlap matching (detected value ⊇ or ≈ expected value),
 * per entity category. Reports Precision / Recall / F1.
 *
 * NOTE: PERSON entities require Mode.SMART (NER). Regex-only tools show 0%
 *       recall for the 'person' category — this demonstrates the NER gap.
 */

import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join, resolve } from "path";
import { performance } from "perf_hooks";

const __dirname = dirname(fileURLToPath(import.meta.url));
const CORPUS_PATH = resolve(__dirname, "../corpus/pii_samples.json");

// ---------------------------------------------------------------------------
// Corpus
// ---------------------------------------------------------------------------

const samples = JSON.parse(readFileSync(CORPUS_PATH, "utf8"));

// ---------------------------------------------------------------------------
// Entity type normalisation
// OpenRedaction uses different names → canonical
// ---------------------------------------------------------------------------

const OR_TYPE_MAP = {
  EMAIL:              "EMAIL",
  PHONE_US:           "PHONE",
  PHONE_UK_MOBILE:    "PHONE",
  PHONE_INTL:         "PHONE",
  SSN:                "SSN",
  CREDIT_CARD_VISA:   "CREDIT_CARD",
  CREDIT_CARD_MASTERCARD: "CREDIT_CARD",
  CREDIT_CARD_AMEX:   "CREDIT_CARD",
  IBAN:               "IBAN",
  IP_ADDRESS:         "IP_ADDRESS",
  URL:                "URL",
  API_KEY:            "API_KEY",
  JWT:                "JWT",
  BEARER_TOKEN:       "JWT",
  NAME:               "PERSON",
};

// prompt-sanitizer JS uses slightly different names
const PS_TYPE_MAP = {
  EMAIL:            "EMAIL",
  PHONE:            "PHONE",
  SSN:              "SSN",
  CREDIT_CARD:      "CREDIT_CARD",
  IBAN:             "IBAN",
  IP_ADDRESS:       "IP_ADDRESS",
  URL:              "URL",
  API_KEY:          "API_KEY",
  JWT_TOKEN:        "JWT",
  PERSON_NAME:      "PERSON",
  SECRET_KEY:       "API_KEY",
  AWS_KEY:          "API_KEY",
};

// ---------------------------------------------------------------------------
// Matching
// ---------------------------------------------------------------------------

function valueMatch(detected, expected) {
  const d = detected.trim().toLowerCase();
  const e = expected.trim().toLowerCase();
  return d === e || e.includes(d) || d.includes(e);
}

function computeStats(detectionsList) {
  const categoryStats = {};
  let totalTP = 0, totalFP = 0, totalFN = 0;

  for (let i = 0; i < samples.length; i++) {
    const sample = samples[i];
    const detected = detectionsList[i] ?? [];
    const expected = sample.entities;
    const cat = sample.category;

    if (!categoryStats[cat]) categoryStats[cat] = { tp: 0, fp: 0, fn: 0 };

    const matchedDetected = new Set();

    for (const exp of expected) {
      let found = false;
      for (let j = 0; j < detected.length; j++) {
        if (matchedDetected.has(j)) continue;
        if (detected[j].type === exp.type && valueMatch(detected[j].value, exp.value)) {
          found = true;
          matchedDetected.add(j);
          break;
        }
      }
      if (found) {
        categoryStats[cat].tp++;
        totalTP++;
      } else {
        categoryStats[cat].fn++;
        totalFN++;
      }
    }

    // FP
    for (let j = 0; j < detected.length; j++) {
      if (!matchedDetected.has(j)) {
        const relevantTypes = new Set(expected.map((e) => e.type));
        if (!expected.length || relevantTypes.has(detected[j].type)) {
          const t = detected[j].type;
          if (!categoryStats[t]) categoryStats[t] = { tp: 0, fp: 0, fn: 0 };
          categoryStats[t].fp++;
          totalFP++;
        }
      }
    }
  }

  categoryStats["OVERALL"] = { tp: totalTP, fp: totalFP, fn: totalFN };
  return categoryStats;
}

function pct(n, d) {
  return d === 0 ? "  N/A" : `${((n / d) * 100).toFixed(1)}%`;
}

function f1(tp, fp, fn) {
  const p = tp / (tp + fp || 1);
  const r = tp / (tp + fn || 1);
  const score = p + r === 0 ? 0 : (2 * p * r) / (p + r);
  return `${(score * 100).toFixed(1)}%`;
}

function printResults(toolName, statsMap) {
  console.log(`\n${"=".repeat(65)}`);
  console.log(`  ${toolName}`);
  console.log("=".repeat(65));
  console.log(`  ${"Category".padEnd(22)} ${"Precision".padStart(10)} ${"Recall".padStart(10)} ${"F1".padStart(8)}   TP  FP  FN`);
  console.log(`  ${"-".repeat(22)} ${"-".repeat(10)} ${"-".repeat(10)} ${"-".repeat(8)}   --  --  --`);

  for (const [cat, s] of Object.entries(statsMap).sort()) {
    const prec = pct(s.tp, s.tp + s.fp);
    const rec  = pct(s.tp, s.tp + s.fn);
    const f    = f1(s.tp, s.fp, s.fn);
    console.log(`  ${cat.padEnd(22)} ${prec.padStart(10)} ${rec.padStart(10)} ${f.padStart(8)}  ${String(s.tp).padStart(3)} ${String(s.fp).padStart(3)} ${String(s.fn).padStart(3)}`);
  }
}

// ---------------------------------------------------------------------------
// prompt-sanitizer runner
// ---------------------------------------------------------------------------

async function runPromptSanitizer(mode) {
  const { Sanitizer, Mode } = await import(
    resolve(__dirname, "../../packages/javascript/src/index.js")
  ).catch(() =>
    import(resolve(__dirname, "../../packages/javascript/dist/index.js"))
  );

  const s = new Sanitizer({ mode: Mode[mode], onDetect: "warn" });
  const t0 = performance.now();
  const detectionsList = [];

  for (const sample of samples) {
    const result = await s.sanitize(sample.text);
    detectionsList.push(
      result.entities.map((e) => ({
        type: PS_TYPE_MAP[e.entityType] ?? e.entityType,
        value: e.value,
      }))
    );
  }

  const avgMs = (performance.now() - t0) / samples.length;
  return { detectionsList, avgMs };
}

// ---------------------------------------------------------------------------
// OpenRedaction runner
// ---------------------------------------------------------------------------

async function runOpenRedaction() {
  let OpenRedaction;
  try {
    ({ OpenRedaction } = await import("openredaction"));
  } catch {
    console.log("\n  [SKIP] openredaction not installed (npm install openredaction)");
    return { detectionsList: samples.map(() => []), avgMs: 0 };
  }

  const redactor = new OpenRedaction({ enableContextAnalysis: true });
  const t0 = performance.now();
  const detectionsList = [];

  for (const sample of samples) {
    const result = await redactor.detect(sample.text);
    const detections = (result.detections ?? []).map((d) => ({
      type: OR_TYPE_MAP[d.type] ?? d.type,
      value: d.value,
    }));
    detectionsList.push(detections);
  }

  const avgMs = (performance.now() - t0) / samples.length;
  return { detectionsList, avgMs };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const total = samples.length;
  const nerSamples = samples.filter((s) => ["person", "multi"].includes(s.category)).length;
  console.log(`\nLoaded ${total} samples (${nerSamples} require NER for full score)`);

  // ── prompt-sanitizer FAST ─────────────────────────────────────────────────
  console.log("\n[1/3] Running prompt-sanitizer (FAST) ...");
  const { detectionsList: psFast, avgMs: psFastMs } = await runPromptSanitizer("FAST");
  const psFastStats = computeStats(psFast);
  printResults(`prompt-sanitizer  FAST  (${psFastMs.toFixed(2)} ms/call avg)`, psFastStats);

  // ── prompt-sanitizer SMART ────────────────────────────────────────────────
  console.log("\n[2/3] Running prompt-sanitizer (SMART — NER) ...");
  const { detectionsList: psSmart, avgMs: psSmartMs } = await runPromptSanitizer("SMART");
  const psSmartStats = computeStats(psSmart);
  printResults(`prompt-sanitizer  SMART (${psSmartMs.toFixed(2)} ms/call avg)`, psSmartStats);

  // ── OpenRedaction ─────────────────────────────────────────────────────────
  console.log("\n[3/3] Running OpenRedaction ...");
  const { detectionsList: orDetections, avgMs: orMs } = await runOpenRedaction();
  const orStats = computeStats(orDetections);
  printResults(`OpenRedaction (regex)   (${orMs.toFixed(2)} ms/call avg)`, orStats);

  // ── Summary ───────────────────────────────────────────────────────────────
  console.log(`\n${"=".repeat(70)}`);
  console.log("  OVERALL SUMMARY");
  console.log("=".repeat(70));
  console.log(`  ${"Tool".padEnd(42)} ${"Precision".padStart(10)} ${"Recall".padStart(10)} ${"F1".padStart(8)}`);
  console.log(`  ${"-".repeat(42)} ${"-".repeat(10)} ${"-".repeat(10)} ${"-".repeat(8)}`);

  for (const [name, statsMap, ms] of [
    [`prompt-sanitizer FAST  (${psFastMs.toFixed(1)}ms/call)`, psFastStats, psFastMs],
    [`prompt-sanitizer SMART (${psSmartMs.toFixed(1)}ms/call)`, psSmartStats, psSmartMs],
    [`OpenRedaction regex    (${orMs.toFixed(1)}ms/call)`, orStats, orMs],
  ]) {
    const s = statsMap["OVERALL"];
    if (!s) continue;
    const prec = pct(s.tp, s.tp + s.fp);
    const rec  = pct(s.tp, s.tp + s.fn);
    const f    = f1(s.tp, s.fp, s.fn);
    console.log(`  ${name.padEnd(42)} ${prec.padStart(10)} ${rec.padStart(10)} ${f.padStart(8)}`);
  }

  console.log();
}

main().catch(console.error);

File: benchmarks/python/run_accuracy.py
"""
Python Accuracy Benchmark — prompt-sanitizer vs Presidio vs LLM Guard
======================================================================

Usage:
    pip install -r requirements.txt
    python run_accuracy.py

    # Skip tools not installed:
    python run_accuracy.py --skip-presidio
    python run_accuracy.py --skip-llmguard

Metric: Value-overlap matching (detected entity value ⊇ or ≈ ground-truth value),
per entity category and overall. Reports Precision / Recall / F1.

NOTE: PERSON entities require NER (Mode.SMART). Regex-only tools will show 0% recall
      for the 'person' category — this is expected and illustrates the gap.
"""
from __future__ import annotations

import argparse
import json
import sys
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

# ---------------------------------------------------------------------------
# Corpus loading
# ---------------------------------------------------------------------------

CORPUS_PATH = Path(__file__).parent.parent / "corpus" / "pii_samples.json"

def load_corpus() -> list[dict]:
    with open(CORPUS_PATH) as f:
        return json.load(f)


# ---------------------------------------------------------------------------
# Entity-type normalisation
# Maps each tool's label → our canonical type (for fair comparison)
# ---------------------------------------------------------------------------

# Presidio → our canonical
PRESIDIO_MAP: dict[str, str] = {
    "EMAIL_ADDRESS":   "EMAIL",
    "PHONE_NUMBER":    "PHONE",
    "US_SSN":          "SSN",
    "CREDIT_CARD":     "CREDIT_CARD",
    "IBAN_CODE":       "IBAN",
    "IP_ADDRESS":      "IP_ADDRESS",
    "URL":             "URL",
    "PERSON":          "PERSON",
    "ORGANIZATION":    "ORGANIZATION",
    "LOCATION":        "LOCATION",
    "CRYPTO":          "CRYPTO_ADDRESS",
    "DATE_TIME":       "DATE",
}

# LLM Guard → our canonical
LLMGUARD_MAP: dict[str, str] = {
    "EMAIL_ADDRESS":   "EMAIL",
    "EMAIL_ADDRESS_RE":"EMAIL",
    "PHONE_NUMBER":    "PHONE",
    "US_SSN":          "SSN",
    "US_SSN_RE":       "SSN",
    "CREDIT_CARD":     "CREDIT_CARD",
    "CREDIT_CARD_RE":  "CREDIT_CARD",
    "IBAN_CODE":       "IBAN",
    "IP_ADDRESS":      "IP_ADDRESS",
    "PERSON":          "PERSON",
    "ORGANIZATION":    "ORGANIZATION",
    "LOCATION":        "LOCATION",
    "CRYPTO":          "CRYPTO_ADDRESS",
    "URL":             "URL",
    "UUID":            "OTHER",
}


# ---------------------------------------------------------------------------
# Scoring helpers
# ---------------------------------------------------------------------------

def _value_match(detected_val: str, expected_val: str) -> bool:
    """True if detected value contains or is contained by the expected value."""
    d = detected_val.strip().lower()
    e = expected_val.strip().lower()
    return d == e or e in d or d in e


@dataclass
class CategoryStats:
    tp: int = 0
    fp: int = 0
    fn: int = 0

    @property
    def precision(self) -> float:
        return self.tp / (self.tp + self.fp) if (self.tp + self.fp) > 0 else 0.0

    @property
    def recall(self) -> float:
        return self.tp / (self.tp + self.fn) if (self.tp + self.fn) > 0 else 0.0

    @property
    def f1(self) -> float:
        p, r = self.precision, self.recall
        return 2 * p * r / (p + r) if (p + r) > 0 else 0.0


def score(
    samples: list[dict],
    detections: list[list[dict]],  # [{"type": canonical_type, "value": str}, ...]
    categories: list[str] | None = None,
) -> dict[str, CategoryStats]:
    """
    Compare detections against ground truth.

    Returns per-category stats plus "overall" key.
    """
    stats: dict[str, CategoryStats] = {}
    totals = CategoryStats()

    for sample, detected in zip(samples, detections):
        cat = sample["category"]
        if categories and cat not in categories:
            continue
        if cat not in stats:
            stats[cat] = CategoryStats()

        expected = sample["entities"]  # list of {type, value}

        # For each expected entity: find a matching detection
        matched_detected = set()
        for exp in expected:
            found = False
            for i, det in enumerate(detected):
                if i in matched_detected:
                    continue
                # type must match at category level (loose: both EMAIL, both PHONE etc.)
                if det["type"] == exp["type"] and _value_match(det["value"], exp["value"]):
                    found = True
                    matched_detected.add(i)
                    break
            if found:
                stats[cat].tp += 1
                totals.tp += 1
            else:
                stats[cat].fn += 1
                totals.fn += 1

        # FP: any detection not matched to expected
        for i, det in enumerate(detected):
            if i not in matched_detected:
                # Only count FP for expected categories (ignore unrelated detections)
                relevant_types = {e["type"] for e in expected}
                if not expected or det["type"] in relevant_types:
                    stats.setdefault(det["type"], CategoryStats()).fp += 1
                    totals.fp += 1

    stats["OVERALL"] = totals
    return stats


def print_results(tool_name: str, stats: dict[str, CategoryStats]) -> None:
    print(f"\n{'='*60}")
    print(f"  {tool_name}")
    print(f"{'='*60}")
    print(f"  {'Category':<20} {'Precision':>10} {'Recall':>10} {'F1':>10}  TP  FP  FN")
    print(f"  {'-'*20} {'-'*10} {'-'*10} {'-'*10}  --  --  --")
    for cat, s in sorted(stats.items()):
        print(
            f"  {cat:<20} {s.precision:>9.1%} {s.recall:>9.1%} {s.f1:>9.1%}"
            f"  {s.tp:2}  {s.fp:2}  {s.fn:2}"
        )


# ---------------------------------------------------------------------------
# prompt-sanitizer runner
# ---------------------------------------------------------------------------

def run_prompt_sanitizer(samples: list[dict]) -> tuple[list[list[dict]], float]:
    """Run prompt-sanitizer FAST mode. Returns (detections_per_sample, avg_ms)."""
    import asyncio
    sys.path.insert(0, str(Path(__file__).parent.parent.parent / "packages" / "python"))
    from prompt_sanitizer import Sanitizer, Mode  # type: ignore

    s = Sanitizer(mode=Mode.FAST, on_detect="warn")

    async def _run_all():
        start = time.perf_counter()
        results = []
        for sample in samples:
            r = await s.sanitize(sample["text"])
            results.append([
                {"type": e.entity_type.value, "value": e.original}
                for e in r.entities
            ])
        elapsed_ms = (time.perf_counter() - start) * 1000
        return results, elapsed_ms / len(samples)

    return asyncio.run(_run_all())


def run_prompt_sanitizer_smart(samples: list[dict]) -> tuple[list[list[dict]], float]:
    """Run prompt-sanitizer SMART mode (NER enabled)."""
    import asyncio
    sys.path.insert(0, str(Path(__file__).parent.parent.parent / "packages" / "python"))
    from prompt_sanitizer import Sanitizer, Mode  # type: ignore

    s = Sanitizer(mode=Mode.SMART, on_detect="warn")

    async def _run_all():
        start = time.perf_counter()
        results = []
        for sample in samples:
            r = await s.sanitize(sample["text"])
            results.append([
                {"type": e.entity_type.value, "value": e.original}
                for e in r.entities
            ])
        elapsed_ms = (time.perf_counter() - start) * 1000
        return results, elapsed_ms / len(samples)

    return asyncio.run(_run_all())


# ---------------------------------------------------------------------------
# Presidio runner
# ---------------------------------------------------------------------------

def run_presidio(samples: list[dict]) -> tuple[list[list[dict]], float]:
    """Run Presidio analyzer. Requires: pip install presidio-analyzer presidio-anonymizer"""
    try:
        from presidio_analyzer import AnalyzerEngine  # type: ignore
    except ImportError:
        print("  [SKIP] presidio-analyzer not installed.")
        return [[] for _ in samples], 0.0

    print("  Loading Presidio AnalyzerEngine (spaCy)...")
    analyzer = AnalyzerEngine()

    start = time.perf_counter()
    results = []
    for sample in samples:
        raw = analyzer.analyze(text=sample["text"], language="en")
        results.append([
            {
                "type": PRESIDIO_MAP.get(r.entity_type, r.entity_type),
                "value": sample["text"][r.start:r.end],
            }
            for r in raw
        ])
    elapsed_ms = (time.perf_counter() - start) * 1000
    return results, elapsed_ms / len(samples)


# ---------------------------------------------------------------------------
# LLM Guard runner
# ---------------------------------------------------------------------------

def run_llmguard(samples: list[dict]) -> tuple[list[list[dict]], float]:
    """
    Run LLM Guard Anonymize scanner.
    Requires: pip install llm-guard

    Note: LLM Guard does not expose individual entity spans/values in its public API.
    We infer detections by comparing the sanitized output against the original:
    any expected PII value absent from the output is counted as detected (recall proxy).
    For precision: any expected value present in the output despite being PII = FP.
    """
    try:
        from llm_guard.vault import Vault           # type: ignore
        from llm_guard.input_scanners import Anonymize  # type: ignore
    except ImportError:
        print("  [SKIP] llm-guard not installed.")
        return [[] for _ in samples], 0.0

    print("  Loading LLM Guard scanner (DeBERTa download on first run)...")
    vault = Vault()
    scanner = Anonymize(vault=vault, use_faker=False)

    start = time.perf_counter()
    results = []
    for sample in samples:
        sanitized, _is_valid, _score = scanner.scan(sample["text"])
        # Infer detections: check which expected values are absent from the output
        detected = []
        for exp in sample["entities"]:
            if exp["value"].lower() not in sanitized.lower():
                detected.append({"type": exp["type"], "value": exp["value"]})
        results.append(detected)
        # Reset vault between samples to avoid cross-contamination
        vault._entries.clear() if hasattr(vault, "_entries") else None

    elapsed_ms = (time.perf_counter() - start) * 1000
    return results, elapsed_ms / len(samples)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(description="PII accuracy benchmark")
    parser.add_argument("--skip-presidio", action="store_true")
    parser.add_argument("--skip-llmguard", action="store_true")
    parser.add_argument("--only", help="comma-separated categories to test")
    args = parser.parse_args()

    samples = load_corpus()
    filter_cats = args.only.split(",") if args.only else None
    if filter_cats:
        samples = [s for s in samples if s["category"] in filter_cats]

    total = len(samples)
    ner_samples = sum(1 for s in samples if s["category"] in ("person", "multi"))
    print(f"\nLoaded {total} samples ({ner_samples} require NER for full score)")

    # ── prompt-sanitizer FAST ─────────────────────────────────────────────────
    print("\n[1/4] Running prompt-sanitizer (FAST mode) ...")
    ps_fast_detections, ps_fast_ms = run_prompt_sanitizer(samples)
    ps_fast_stats = score(samples, ps_fast_detections)
    print_results(f"prompt-sanitizer  FAST  ({ps_fast_ms:.2f} ms/call avg)", ps_fast_stats)

    # ── prompt-sanitizer SMART ────────────────────────────────────────────────
    print("\n[2/4] Running prompt-sanitizer (SMART mode — NER) ...")
    ps_smart_detections, ps_smart_ms = run_prompt_sanitizer_smart(samples)
    ps_smart_stats = score(samples, ps_smart_detections)
    print_results(f"prompt-sanitizer  SMART ({ps_smart_ms:.2f} ms/call avg)", ps_smart_stats)

    # ── Presidio ──────────────────────────────────────────────────────────────
    if not args.skip_presidio:
        print("\n[3/4] Running Presidio ...")
        pr_detections, pr_ms = run_presidio(samples)
        pr_stats = score(samples, pr_detections)
        print_results(f"Presidio (spaCy)        ({pr_ms:.2f} ms/call avg)", pr_stats)
    else:
        print("\n[3/4] Presidio — SKIPPED")

    # ── LLM Guard ─────────────────────────────────────────────────────────────
    if not args.skip_llmguard:
        print("\n[4/4] Running LLM Guard ...")
        lg_detections, lg_ms = run_llmguard(samples)
        lg_stats = score(samples, lg_detections)
        print_results(f"LLM Guard (DeBERTa)     ({lg_ms:.2f} ms/call avg)", lg_stats)
    else:
        print("\n[4/4] LLM Guard — SKIPPED")

    # ── Comparison summary ────────────────────────────────────────────────────
    print(f"\n{'='*70}")
    print("  OVERALL SUMMARY")
    print(f"{'='*70}")
    print(f"  {'Tool':<40} {'Precision':>10} {'Recall':>10} {'F1':>10}")
    print(f"  {'-'*40} {'-'*10} {'-'*10} {'-'*10}")

    rows = [
        (f"prompt-sanitizer FAST  ({ps_fast_ms:.1f}ms/call)", ps_fast_stats.get("OVERALL", CategoryStats())),
        (f"prompt-sanitizer SMART ({ps_smart_ms:.1f}ms/call)", ps_smart_stats.get("OVERALL", CategoryStats())),
    ]
    if not args.skip_presidio and "pr_stats" in dir():
        rows.append((f"Presidio spaCy         ({pr_ms:.1f}ms/call)", pr_stats.get("OVERALL", CategoryStats())))  # type: ignore[reportPossiblyUnbound]
    if not args.skip_llmguard and "lg_stats" in dir():
        rows.append((f"LLM Guard DeBERTa      ({lg_ms:.1f}ms/call)", lg_stats.get("OVERALL", CategoryStats())))  # type: ignore[reportPossiblyUnbound]

    for name, s in rows:
        print(f"  {name:<40} {s.precision:>9.1%} {s.recall:>9.1%} {s.f1:>9.1%}")

    print()


if __name__ == "__main__":
    main()
